primefaces/src/main/java/org/primefaces/util/ResourceUtils.java

```diff
@@ -335,4 +320,197 @@ public static Resource evaluateResourceExpression(FacesContext facesContext, Val
+     *
+     * @param context The FacesContext
+     * @param libraryName The library name containing the resource
+     * @param resourceName The name of the JavaScript resource
+     */
+    public static void addJavascriptResource(FacesContext context, String libraryName, String resourceName) {
+        addResource(context, RENDERER_SCRIPT, libraryName, resourceName);
+    }
+    /**
+     * Adds a JavaScript resource from the default PrimeFaces library to the view.
+     *
+     * @param context The FacesContext
+     * @param resourceName The name of the JavaScript resource
+     */
+    public static void addJavascriptResource(FacesContext context, String resourceName) {
+        addResource(context, RENDERER_SCRIPT, Constants.LIBRARY, resourceName);
+    }
+
+    /**
+     * Adds a CSS resource to the view.
+     *
+     * @param context The FacesContext
+     * @param libraryName The library name containing the resource
+     * @param resourceName The name of the CSS resource
+     */
+    public static void addStyleSheetResource(FacesContext context, String libraryName, String resourceName) {
+        addResource(context, RENDERER_STYLESHEET, libraryName, resourceName);
+    }
+
+    /**
+     * Adds a CSS resource from the default PrimeFaces library to the view.
+     *
+     * @param context The FacesContext
+     * @param resourceName The name of the CSS resource
+     */
+    public static void addStyleSheetResource(FacesContext context, String resourceName) {
+        addResource(context, RENDERER_STYLESHEET, Constants.LIBRARY, resourceName);
+    }
+
+    /**
+     * Adds a resource to the view if it hasn't been rendered yet.
+     *
+     * @param context The FacesContext
+     * @param type The renderer type (CSS or JS)
+     * @param libraryName The library name containing the resource
+     * @param resourceName The name of the resource
+     */
+    public static void addResource(FacesContext context, String type, String libraryName, String resourceName) {
+        if (!context.getApplication().getResourceHandler().isResourceRendered(context, resourceName, libraryName)) {
+            addResourceToHead(context, type, libraryName, resourceName);
+        }
+    }
+
+    /**
+     * Adds a resource component to the specified target, avoiding duplicates based on ID.
+     *
+     * @param context The FacesContext
+     * @param type The renderer type
+     * @param libraryName The library name
+     * @param resourceName The resource name
+     * @param target The target location ("head" or "body")
+     * @return The added or existing UIComponent
+     */
+    public static UIComponent addScriptResourceToTarget(FacesContext context, String type, String libraryName, String resourceName, String target) {
+        String id = (libraryName != null ? libraryName.replaceAll("\\W+", "_") + "_" : "") + resourceName.replaceAll("\\W+", "_");
+
+        // Check for existing resource with same ID
```

Comment:
i would really like to check JSF impl internals, if we really add a resource a second time there, if its already in the list
i dont think so, otherwise for me its a impl bug

Reply:
This was also in OmniFaces and it may not be necessary anymore but it didn't seem to hurt?

Reply:
im just not happy to move a hack from one place to another place

IMO we should check first aht both impls are NOT adding the component again when we call:
context.getViewRoot().addComponentResource(context, resource, target);
without given the component a ID

Reply:
I will test it.

Reply:
We can add Integration Test for it, it runs on different versions